### PR TITLE
Prohibit patterns in trait methods without bodies

### PR DESCRIPTION
**Pull request by petrochenkov**
They are not properly type checked

``` rust
trait Tr {
    fn f(&a: u8); // <- This compiles
}
```

, mostly rejected by the parser already and generally don't make much sense.
This PR is kind of a missing part of https://github.com/rust-lang/rust/pull/35015.

Given the [statistics from crater](https://github.com/rust-lang/rust/pull/37378#issuecomment-256154994), the effect of this PR is mostly equivalent to improving `unused_mut` lint.

cc https://github.com/rust-lang/rust/issues/35078#issuecomment-255707355 https://github.com/rust-lang/rust/pull/35015 https://github.com/rust-lang/rfcs/pull/1685 https://github.com/rust-lang/rust/issues/35203
r? @eddyb

### CODE REVIEW

**Comment by eddyb** [comment]
Started crater run.

**Comment by eddyb** [comment]
[Crater run](https://gist.github.com/eddyb/60128b937fd2bc4570234d1f7af77bd0) looks pretty bad (177 root regressions).

**Comment by petrochenkov** [comment]
Something is wrong with root/non-root partitioning in crater.
Two very popular crates  - `rustc-serialize` and `num-traits` have this problem, most of others seem to break because of these dependencies.
I turned the error into a deny-by-default lint, @eddyb can you run crater once more to determine the real number of root regressions? (Unless you think it's waste of resources.)

Anyway, it looks like we can't make this an error in the nearest future, but I still hope to proceed with a warn-by-default lint.

**Comment by eddyb** [comment]
@petrochenkov Maybe they don't show themselves? They might be broken or missing?

**Comment by eddyb** [comment]
Second [crater report](https://gist.github.com/eddyb/18c471b140e8339c8de70e0f907a8fa4) shows only 22 root regressions.

**Comment by petrochenkov** [comment]
`mut arg` - 19 crates (if only `unused_mut` worked in this context...).
`&arg` - 3 crates.

I've updated the PR with warning-by-default and tracking issue number.
Also https://github.com/rust-lang/rust/issues/35203 needs to be reopened (I have no rights to do it).

**Comment by Manishearth** [comment]
Fails tidy, https://github.com/rust-lang/rust/pull/37416/commits/d230a7ecc3a8c5dd16a953d3331cbf91c8b44b8f

**Comment by TimNN** [comment]
The used error code (`E0570`) has been merged by a different PR  (#36421) since this PR was opened (thus this PR currently fails travis when merged).

File: src/librustc_passes/diagnostics.rs [hunk 1]
```diff
@@ -228,4 +228,5 @@ pub impl Foo for Bar {
 register_diagnostics! {
     E0472, // asm! is unsupported on this target
     E0561, // patterns aren't allowed in function pointer types
+    E0570, // patterns aren't allowed in methods without bodies
```
Comment (Manishearth) [inline src/librustc_passes/diagnostics.rs]: This seems to be unused?

**Comment by retep998** [comment]
I believe this caused https://github.com/rust-lang/rust/pull/37416 to fail

https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2155

```
error: patterns aren't allowed in methods without bodies
   --> tests\cargotest\support/mod.rs:604:34
    |
604 |     fn tap<F: FnOnce(&mut Self)>(mut self, callback: F) -> Self;
    |                                  ^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #35203 <https://github.com/rust-lang/rust/issues/35203>
note: lint level defined here
   --> tests\cargotest\lib.rs:1:9
    |
1   | #![deny(warnings)]
    |         ^^^^^^^^

error: aborting due to previous error
```

**Comment by TimNN** [comment]
This probably needs to update the sha at https://github.com/rust-lang/rust/blob/master/src/tools/cargotest/main.rs#L27 to at  least rust-lang/cargo@806e3c368a15f618244a3b4e918bf77f9c403fd0

**Comment by WaDelma** [comment]
Hey that crater run that said 22 root regressions didn't fail to build my [crate](https://crates.io/crates/poisson) even though it quite few traits with anonymous parameters. Was the lint broken or is there something I am missing?

**Comment by petrochenkov** [comment]
@WaDelma 
This lint reports patterns in parameters of trait methods without body 
```
fn f(&arg: Type);
fn g(mut arg: Type);
```
, anonymous parameters
```
fn h(Type);
```
are not reported (yet, waiting for https://github.com/rust-lang/rfcs/pull/1685).

**Comment by WaDelma** [comment]
Ah okay. That was the part I was missing. :D

**Pull request merged**